### DISCUSSION TITLE
Training Thai transformer & tokenizer

### DISCUSSION THREAD

**Discussion by kanayer** [explosion/spaCy#10542] [Help: Coding & Implementations]
I am trying to train the spaCy pipeline for the Thai language consisting of a transformer, tagger, and parser and I was wondering which model is the best for the task. I have cloned the [UD Benchmark](https://explosion.ai/blog/ud-benchmarks-v3-2) pipeline and removed the `experimental character-based NER tokenizer` and the `experimental edit tree lemmatizer `, basically leaving only the transformer component. My dataset consists of `10 000` sentences with words that were segmented properly.

As you know, the default transformer in the UD Benchmark pipeline is `xlm-roberta-base`. When I train it on my dataset together with the default tokenizer `spacy.Tokenizer.v1`, I run into the following error message on the `spacy project run evaluate` step:

> Token indices sequence length is longer than the specified maximum sequence length for this model (642 > 512). Running this sequence through the model will result in indexing errors. 

The sentence length in the dataset varies from 1 to 33 words. Could this be a problem? Is it possible to edit the maximum sequence length for the `xlm-roberta-base` model?

Also, [training quickstart](https://spacy.io/usage/training) recommends `monsoon-nlp/bert-base-thai` as the language-specific transformer model. However, there is the following warning message on the [HuggingFace page](https://huggingface.co/monsoon-nlp/bert-base-thai) of this transformer model:

> You must run the original ThaiTokenizer to have your tokenization match that of the original model.

Is it possible to use `ThaiTokenizer` instead of the `experimental character-based NER tokenizer` in the spaCy pipeline? 
And overall, which transformer model would you advise training the Thai dataset (with properly segmented words) on?

**Comment by polm** [comment]
> The sentence length in the dataset varies from 1 to 33 words. Could this be a problem? Is it possible to edit the maximum sequence length for the xlm-roberta-base model?

I'm not exactly sure what would cause that, but I think what is happening is that because your tokenizer doesn't match the one used for xml-roberta-base you're getting OOV tokens or something. Arbitrarily long sequences should be handled by [striding](https://spacy.io/api/transformer/#span_getters); it's possible something is off with your striding setting. xml-roberta-base probably uses a simple tokenizer that will give different results than properly tokenized or character tokenized Thai.

> Is it possible to use ThaiTokenizer instead of the experimental character-based NER tokenizer in the spaCy pipeline?

Yes, it looks like ThaiTokenizer is the default tokenizer in our Thai configuration. I don't think any of us are very familiar with it, so you'd have to confirm the settings were the same as whatever model you were using, but it should be similar.

> And overall, which transformer model would you advise training the Thai dataset (with properly segmented words) on?

The model in the quickstart for Thai, like many languages we don't provide full pipelines for, is one that seemed popular and that we confirmed would run in spaCy. However we didn't confirm the quality of results or compare it with other models, so we can't really give you any recommendation there.

**Reply by kanayer** [reply]
if ThaiTokenizer is the default tokenizer, how can I call it in the config file? Right now I call `spacy.Tokenizer.v1`, is there a way to call `ThaiTokenizer`?

**Reply by polm** [reply]
You can use `spacy.th.ThaiTokenizer`, or if you remove your tokenizer specification in the config it'll be filled in automatically, since it's a default. Look at the [source for the Thai Language class](https://github.com/explosion/spaCy/blob/c17980e535a8009b14ee4d1f818db207d9c07e55/spacy/lang/th/__init__.py#L13).

**Reply by adrianeboyd** [reply]
The `spacy-transformers` options at this point don't really support custom tokenizers on the transformers side of things, so I would recommend choosing another model. (It's not entirely impossible in general, but it requires a fair amount of work to develop custom transformer components.)

**Reply by kanayer** [reply]
Thank you for your answer! Should `spacy.th.ThaiTokenizer` be trained before the transformer as a separate component? When I used it directly in the transformer config file together with the `monsoon-nlp/bert-base-thai` transformer, the accuracy of the model was very low. 

Since `spacy.th.ThaiTokenizer` uses `pythainlp` library, will the accuracy be still low if the way `pythainlp` tokenizes words is different from my dataset? In such case, would it be better to use the `spacy.Tokenizer.v1` with a different transformer? 

Do you have any other model than `xml-roberta-base` on mind that can be used with Thai language?

**Reply by adrianeboyd** [reply]
None of the tokenizers in `[nlp.tokenizer]` are trainable from within spacy. (You can train things like pkuseg models outside of spacy, of course.) Within spacy you can only train things that are in `[nlp.pipeline]`.

The problem is that the spacy tokens aren't passed to the transformer tokenizer in `spacy-transformers`, just the text. You would need custom components to handle a case like for this Thai transformer model. (So for the transformer it doesn't matter what tokenizer is in your spacy config. The two tokenizations are kept completely separate and just aligned after processing with `spacy-alignments`.)

Sorry, I don't think we don't have the expertise to recommend any other models for Thai.

**Reply by kanayer** [reply]
Okay, understood, thank you. If instead of spacy tokens, only text is passed to the transformer tokenizer, does the metadata `"#text"` affect the transformer tokenization? The thing is I have non-segmented text in `#text` and properly segmented text in `#segmented` as shown below:

<img width="1035" alt="Screen Shot 2022-03-25 at 1 29 53 PM" src="https://user-images.githubusercontent.com/22777876/160055035-bb338c32-6c40-4ffc-96ab-ab37d3742593.png">

When I train this dataset using the `xml-roberta-base` transformer, the training accuracy is very good however, there is an error in the evaluation process as I mentioned earlier:

> Token indices sequence length is longer than the specified maximum sequence length for this model (642 > 512). Running this sequence through the model will result in indexing errors.

As @polm explained, the error above happens because my tokenizer doesn't match the one used for `xml-roberta-base`. The tokenizer I used was `spacy.Tokenizer.v1` and it works perfectly in the statistical model. I was wondering is it possible/a good idea to disable transformer's tokenizer? Or would changing the model help given that most transformers have a maximum sequence length of 512 tokens?

**Reply by adrianeboyd** [reply]
About the transformer tokenizer length warning: https://github.com/explosion/spaCy/discussions/9277#discussioncomment-1374226

The spacy CoNLL-U converter uses `SpaceAfter=No` to handle spacing, not `# text`, so the example above won't be converted correctly.

**Reply by kanayer** [reply]
Oh! Then do I understand correctly that if I add `SpaceAfter=No` and `SpaceAfter=Yes` to the dataset, it should handle spacing properly? Also, if I want `SpaceAfter` to represent text in #segmented, do I have to rename it to `#text`?

I'm just curious, could you please briefly explain how CoNLL-U converter affects tokenizer?

**Reply by adrianeboyd** [reply]
There is no `SpaceAfter=Yes`, just `SpaceAfter=No`.

The spacy CoNLL-U converter currently ignores all lines that start with `#`.

The converter and the tokenizer are not connected in any way. The converter does not use a tokenizer and how you converted the data does not affect the tokenizer in your config.

**Reply by kanayer** [reply]
Oh, sorry for the stupid question but then what did you mean by this? 
> The spacy CoNLL-U converter uses SpaceAfter=No to handle spacing, not # text, the example above won't be converted correctly.

If I need tokens to be converted as in `#segmented`, and all # are ignored, then no need to add `SpaceAfter` because it is already space separated as in `#segmented`? 

And since `SpaceAfter` has no effect on tokenizer, the only thing that might help resolve the problem of the error about `maximum sequence length` is lowering `window` and `stride`?

**Reply by adrianeboyd** [reply]
With the spacy converter you can control the trailing spaces in the converted docs by whether you have `SpaceAfter=No` in `MISC`. Just check the docs in `output.spacy` to see whether the spacing is what you want.

And I guess what I said was potentially a bit misleading: whether the text itself contains whitespace between tokens will obviously affect what comes out of the tokenizer. However, the underlying tokenizer config and how it handles (a lack of) whitespace isn't affected by how the docs were converted. You have to configure your tokenizer so it works for your particular input data.
